### ISSUE TITLE
Improving strategy docs

### ISSUE THREAD

**Issue by rsokl** [HypothesisWorks/hypothesis#2957]
While working on a tutorial for pycon, I found that a substantial part of my materials was simply explaining what the various strategies do... which is what docstrings are for, but I think that ours fall short.

I am a fan of [NumPy-style docs](https://numpy.org/doc/1.20/reference/generated/numpy.all.html#numpy.all) where each function's parameters are clearly described – alongside their type annotations – and an example section follows. Among these qualities, I think including an examples section for each docstring has the biggest bang for its buck.

One issue with this is that the [Core Strategies](https://hypothesis.readthedocs.io/en/latest/data.htm) page may end up being pretty lengthy. Perhaps expanding the TOC to include the strategies would make it easier for users to navigate. These would be collapsable, so this wouldn't blow up the navigation pane.

**Comment by Zac-HD** [comment]
- I'd support adopting Numpy-style structured docstrings for our whole API, though examples are probably most useful for strategies.  If we're doing this we should also investigate linters which can check for malformed docstrings.
- Too many strategies to add them *all* to the table of contents, but arranging them into sections for scalars, collections, etc ([like in our tutorial](https://docs.google.com/presentation/d/1OQLNah25gu_ElpUP0vv379dBDBZXh5LmK6fbnmMJYxM/)) would probably be useful.

**Comment by rsokl** [comment]
> but arranging them into sections for scalars, collections, etc (like in our tutorial) would probably be useful

I had been thinking this as well, but wasn't sure if restructuring the page itself was on the table. I think this is a great idea and makes it easier to get a bird's eye view of all the strategies

**Comment by Zac-HD** [comment]
Anything that doesn't break URLs is on the table 😁

(I've wanted to restructure to follow [the four kinds of documentation](https://documentation.divio.com/) for years, but doing it well - including redirects - is a lot of work and not at the top of my todo list.  I'd be delighted if someone volunteered to do it though!)

**Comment by rsokl** [comment]
> Anything that doesn't break URLs is on the table

That is what I was worried about when I suggested expanding the TOC rather than rearrange the strats. Rearranging strategies into scalars, collections, etc. would break URLs, right? E.g. in the tutorial I link to https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.fixed_dictionaries

**Comment by Zac-HD** [comment]
Yeah, we can rearrange them on the page and add sidebar-linkable subheadings though (which also preserves ctrl-f searchability).

**Comment by epgui** [comment]
FWIW, this format of documentation is very easy to use for libraries that are essentially just a big collection of utility functions: [https://ramdajs.com/docs/](https://ramdajs.com/docs/)

**Comment by Zac-HD** [comment]
redundant with a note in #4309 now
